File: 剑指offer/14.1.剪绳子.py
class Solution:
    def cuttingRope(self, n: int) -> int:
        dp = [0] * (n + 1)

        for i in range(2, n + 1):
            for j in range(1, i):
                dp[i] = max(dp[i], i * (i - j), i * dp[i - j])

        return dp[-1]

File: 剑指offer/14.1.剪绳子.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int cuttingRope(int n)
    {
        vector<int> dp(n + 1, 0);
        dp[1] = 1;
        dp[2] = 1;

        for (int i = 2; i <= n; i++) {
            for (int j = 1; j < i; j++) {
                dp[i] = max(dp[i], max(j * (i - j), j * dp[i - j]));
            }
        }
        return dp.back();
    }
};